src/rustup-cli/rustup_mode.rs

```diff
@@ -1038,3 +1098,52 @@ fn set_default_host_triple(cfg: &Cfg, m: &ArgMatches<'_>) -> Result<()> {
     cfg.set_default_host_triple(m.value_of("host_triple").expect(""))?;
     Ok(())
 }
+
+#[derive(Copy, Clone, Debug)]
+enum CompletionCommand {
+    Rustup,
+    Cargo,
+}
+
+impl CompletionCommand {
+    fn variants() -> [&'static str; 2] {
+        ["rustup", "cargo"]
+    }
+}
+
+impl FromStr for CompletionCommand {
+    type Err = String;
+
+    fn from_str(s: &str) -> std::result::Result<Self, Self::Err> {
+        match s {
+            _ if s.eq_ignore_ascii_case("rustup") => Ok(CompletionCommand::Rustup),
+            _ if s.eq_ignore_ascii_case("cargo") => Ok(CompletionCommand::Cargo),
+            _ => Err(String::from("[valid values: rustup, cargo]")),
+        }
+    }
+}
+
+fn output_completion_script(shell: Shell, command: CompletionCommand) -> Result<()> {
+    match command {
+        CompletionCommand::Rustup => {
+            cli().gen_completions_to("rustup", shell, &mut io::stdout());
+        }
+        CompletionCommand::Cargo => {
+            let script = match shell {
+                Shell::Bash => Some("/etc/bash_completion.d/cargo"),
+                Shell::Zsh => Some("/share/zsh/site-functions/_cargo"),
```

Comment:
What about other shells, does `cargo` only support bash and zsh?  If so, should we report an error here too if a different shell is selected?

Reply:
`cargo` currently only has completion scripts for `bash` and `zsh`. I agree that we should report an error if a user specifies a shell that is not yet supported. I will add a message printed out on stderr.